### PR TITLE
Optimize async execution

### PR DESCRIPTION
**Pull request by alexandru**
Unless I'm missing something, there's no reason to protect *every single async call* with `onceOnly` in a `flatMap` chain, what we need here is to only protect the source (i.e. the user provided code), because it's not us (the library authors) that will make that mistake.

This PR takes the `onceOnly` protection from `unsafeRunAsync` and puts it where it belongs, in `IO.async`. This cuts down on the number of `getAndSet` operations in a `flatMap` loop.

### CODE REVIEW

**Comment by djspiewak** [comment]
I'm relatively certain this breaks referential transparency.  Something like this should reveal the problem:

```scala
val ioa: IO[Unit] = IO.async(_(Right(())))

(ioa >> ioa).unsafeRunSync()
```

That should hang with this implementation.

**Comment by alexandru** [comment]
@djspiewak I don't understand, why should it hang? I don't think so - I just tested it.

Also:

```scala
val io = IO(10)
(for (a <- io; b <- io) yield a + b).unsafeRunSync

//=> res: Int = 20
```

**Comment by alexandru** [comment]
I can add a test for it if you want.

**Comment by djspiewak** [comment]
Oh, you're right!  Because it's inside the registration, and we discard that context.  Adding a test would be very good, because this is a relatively easy thing to break, but I agree that it's not broken in its current form.

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -533,13 +533,13 @@ object IO extends IOInstances {
...
-   * IO.fromFuture(f).unsafeToFuture === f     // true-ish (except for memoization)
-   * IO.fromFuture(ioa.unsafeToFuture) === ioa // true!
+   * IO.deferFuture(f).unsafeToFuture === f     // true-ish (except for memoization)
+   * IO.deferFuture(ioa.unsafeToFuture) === ioa // true!
    * }}}
    *
    * @see [[IO#unsafeToFuture]]
    */
-  def fromFuture[A](f: => Future[A])(implicit ec: ExecutionContext): IO[A] = {
+  def deferFuture[A](f: => Future[A])(implicit ec: ExecutionContext): IO[A] = {
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IO.scala]: Missed this in my earlier read-over.  I really prefer `fromFuture` to `deferFuture`.  If other people disagree though, I will yield.

**Comment by alexandru** [comment]
Shit, I pushed by mistake another proposal I wanted to make ... renaming `fromFuture` to `deferFuture`.

Will fix.

File: laws/shared/src/test/scala/cats/effect/IOTests.scala [hunk 2]
```diff
@@ -162,6 +165,42 @@ class IOTests extends BaseTestsSuite {
...
+    val io = IO.async[Int] { _ => throw dummy }
+    val f = io.unsafeToFuture()
+
+    ec.tick()
+    f.value shouldEqual Some(Failure(dummy))
+  }
+
+  testAsync("IO.async does not break referential transparency") { implicit ec =>
+    val io = IO.async[Int](_(Right(10)))
+    val sum = for (a <- io; b <- io; c <- io) yield a + b + c
```
Comment (djspiewak) [inline laws/shared/src/test/scala/cats/effect/IOTests.scala]: 👍

**Comment by alexandru** [comment]
OK, something was messed up with this branch, but I fixed it, should be now good to merge :-)

**Pull request merged**